server/src/main/java/io/crate/execution/ddl/tables/AddColumnTask.java

```diff
@@ -82,24 +83,29 @@ public ClusterState execute(ClusterState currentState, AddColumnRequest request)
         DocTableInfoFactory docTableInfoFactory = new DocTableInfoFactory(nodeContext);
         DocTableInfo currentTable = docTableInfoFactory.create(request.relationName(), currentState);
 
+        request = addMissingParentColumns(request, currentTable);
+
+        long currentOID = currentState.metadata().columnOID();
         boolean allExist = true;
-        for (Reference ref : request.references()) {
-            Reference exists = currentTable.getReference(ref.column());
-            if (exists == null) {
+        for (int i = 0; i < request.references().size(); i++) {
+            Reference existingRef = currentTable.getReference(request.references().get(i).column());
+            var refWithOid = tryToAssignOid(existingRef, request.references().get(i), currentOID);
+            if (refWithOid != null) {
                 allExist = false;
-            } else if (ref.valueType().id() != exists.valueType().id()) {
-                throw new IllegalArgumentException(String.format(Locale.ENGLISH,
-                    "Column `%s` already exists with type `%s`. Cannot add same column with type `%s`",
-                    ref.column(),
-                    exists.valueType().getName(),
-                    ref.valueType().getName()));
+                request.references().set(i, refWithOid);
+                currentOID = refWithOid.oid(); // Reference which got last assigned OID has the actual value
+            } else {
+                request.references().set(i, existingRef);
```

Comment:
This is needed to protect existing Reference to be overwritten by Reference created by `alter table add column` plan. Planner Reference has unassigned OID and since we build mapping based on request references, we need to replace request ref by existing one. 

Normally add column is not adding existing reference but when mixed with other concurrent insert/update/add column this can be an issue - `DynamicMappingUpdateItest` used to fail without this.